// @dbml/cli/lib/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys$1 = __getOwnPropNames(from), i = 0, n = keys$1.length, key; i < n; i++) {
		key = keys$1[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
let commander = require("commander");
commander = __toESM(commander);
let path = require("path");
path = __toESM(path);
let __dbml_connector = require("@dbml/connector");
let __dbml_core = require("@dbml/core");
let chalk = require("chalk");
chalk = __toESM(chalk);
let figures = require("figures");
figures = __toESM(figures);
let winston = require("winston");
let fs = require("fs");
fs = __toESM(fs);
let __dbml_parse = require("@dbml/parse");
let node_fs = require("node:fs");
const $schema = "https://json.schemastore.org/package";
const name = "@dbml/cli";
const version = "8.2.1";
const description = "";
const main = "lib/index.js";
const license = "Apache-2.0";
const scripts = {
	"test": "vitest run",
	"test:watch": "vitest watch",
	"coverage": "vitest --coverage",
	"build": "rm -rf ./lib && tsc --noEmit && vite build",
	"dev": "vite build --watch",
	"prepublish": "npm run build",
	"lint": "eslint .",
	"lint:fix": "eslint --fix ."
};
const publishConfig = { "access": "public" };
const bin = {
	"dbml2sql": "bin/dbml2sql.js",
	"sql2dbml": "bin/sql2dbml.js",
	"db2dbml": "bin/db2dbml.js"
};
const author = "Holistics <dev@holistics.io>";
const homepage = "https://dbml.dbdiagram.io";
const repository = "https://github.com/holistics/dbml/tree/master/packages/dbml-cli";
const keywords = ["dbml", "dbml-cli"];
const dependencies = {
	"@babel/cli": "^7.21.0",
	"@dbml/connector": "^8.2.1",
	"@dbml/core": "^8.2.1",
	"@dbml/parse": "^8.2.1",
	"bluebird": "^3.5.5",
	"chalk": "^2.4.2",
	"commander": "^2.20.0",
	"esm": "^3.2.25",
	"figures": "^3.2.0",
	"lodash-es": "^4.17.15",
	"pegjs-require-import": "^0.0.2",
	"strip-ansi": "^5.2.0",
	"winston": "^3.2.1"
};
const gitHead = "e60008b58f86f40d60dcf97b3839897b07598445";
const engines = { "node": ">=18" };
var package_default = {
	$schema,
	name,
	version,
	description: "",
	main,
	license,
	scripts,
	publishConfig,
	bin,
	author,
	homepage,
	repository,
	keywords,
	dependencies,
	gitHead,
	engines
};
var { combine, timestamp, printf } = winston.format;
var consoleFormat = printf((info) => {
	const { level, message } = info;
	return `  ${chalk.default.red(level.toUpperCase())}: ${message}\n
  A complete log can be found in:
     ${path.default.resolve(process.cwd(), "dbml-error.log")}`;
});
var fileFormat = printf((info) => {
	const { timestamp: ts, message, stack, rootError } = info;
	let logContent = `${ts}\n${stack ?? message ?? ""}\n`;
	if (rootError) {
		logContent += "\nROOT_ERROR:";
		logContent += `\n${rootError.stack}`;
		if (rootError.location) logContent += `\n${JSON.stringify(rootError.location)}`;
		logContent += "\n";
	}
	return logContent;
});
const consoleLogger = (0, winston.createLogger)({
	format: combine(consoleFormat),
	transports: [new winston.transports.Console({ level: "error" })]
});
const fileLogger = (0, winston.createLogger)({
	format: combine(timestamp(), winston.format.uncolorize(), fileFormat),
	transports: [new winston.transports.File({
		filename: "dbml-error.log",
		level: "error"
	})]
});
var OutputConsolePlugin = class {
	static write(content) {
		console.log(content);
	}
};
var outputConsolePlugin_default = OutputConsolePlugin;
var OutputFilePlugin = class {
	constructor(filePath, header) {
		this.filePath = filePath;
		this.header = header;
		this.isWrite = false;
	}
	start() {
		fs.default.writeFileSync(this.filePath, "");
		this.stream = fs.default.createWriteStream(this.filePath, { flags: "a" });
		if (this.header) this.stream.write(this.header);
		this.isWrite = true;
	}
	write(content) {
		if (!this.isWrite) this.start();
		this.stream.write(content);
	}
};
var outputFilePlugin_default = OutputFilePlugin;
var freeGlobal = typeof global == "object" && global && global.Object === Object && global;
var _freeGlobal_default = freeGlobal;
var freeSelf = typeof self == "object" && self && self.Object === Object && self;
var root = _freeGlobal_default || freeSelf || Function("return this")();
var _root_default = root;
var Symbol$1 = _root_default.Symbol;
var _Symbol_default = Symbol$1;
var objectProto$3 = Object.prototype;
var hasOwnProperty$9 = objectProto$3.hasOwnProperty;
var nativeObjectToString$1 = objectProto$3.toString;
var symToStringTag$1 = _Symbol_default ? _Symbol_default.toStringTag : void 0;
function getRawTag(value) {
	var isOwn = hasOwnProperty$9.call(value, symToStringTag$1), tag = value[symToStringTag$1];
	try {
		value[symToStringTag$1] = void 0;
		var unmasked = true;
	} catch (e) {}
	var result = nativeObjectToString$1.call(value);
	if (unmasked) if (isOwn) value[symToStringTag$1] = tag;
	else delete value[symToStringTag$1];
	return result;
}
var _getRawTag_default = getRawTag;
var nativeObjectToString = Object.prototype.toString;
function objectToString(value) {
	return nativeObjectToString.call(value);
}
var _objectToString_default = objectToString;
var nullTag = "[object Null]", undefinedTag = "[object Undefined]";
var symToStringTag = _Symbol_default ? _Symbol_default.toStringTag : void 0;
function baseGetTag(value) {
	if (value == null) return value === void 0 ? undefinedTag : nullTag;
	return symToStringTag && symToStringTag in Object(value) ? _getRawTag_default(value) : _objectToString_default(value);
}
var _baseGetTag_default = baseGetTag;
function isObjectLike(value) {
	return value != null && typeof value == "object";
}
var isObjectLike_default = isObjectLike;
var symbolTag$1 = "[object Symbol]";
function isSymbol(value) {
	return typeof value == "symbol" || isObjectLike_default(value) && _baseGetTag_default(value) == symbolTag$1;
}
var isSymbol_default = isSymbol;
function arrayMap(array, iteratee) {
	var index = -1, length = array == null ? 0 : array.length, result = Array(length);
	while (++index < length) result[index] = iteratee(array[index], index, array);
	return result;
}
var _arrayMap_default = arrayMap;
var isArray = Array.isArray;
var isArray_default = isArray;
var INFINITY$1 = Infinity;
var symbolProto$1 = _Symbol_default ? _Symbol_default.prototype : void 0, symbolToString = symbolProto$1 ? symbolProto$1.toString : void 0;
function baseToString(value) {
	if (typeof value == "string") return value;
	if (isArray_default(value)) return _arrayMap_default(value, baseToString) + "";
	if (isSymbol_default(value)) return symbolToString ? symbolToString.call(value) : "";
	var result = value + "";
	return result == "0" && 1 / value == -INFINITY$1 ? "-0" : result;
}
var _baseToString_default = baseToString;
function isObject(value) {
	var type = typeof value;
	return value != null && (type == "object" || type == "function");
}
var isObject_default = isObject;
function identity(value) {
	return value;
}
var identity_default = identity;
var asyncTag = "[object AsyncFunction]", funcTag$1 = "[object Function]", genTag = "[object GeneratorFunction]", proxyTag = "[object Proxy]";
function isFunction(value) {
	if (!isObject_default(value)) return false;
	var tag = _baseGetTag_default(value);
	return tag == funcTag$1 || tag == genTag || tag == asyncTag || tag == proxyTag;
}
var isFunction_default = isFunction;
var coreJsData = _root_default["__core-js_shared__"];
var _coreJsData_default = coreJsData;
var maskSrcKey = function() {
	var uid = /[^.]+$/.exec(_coreJsData_default && _coreJsData_default.keys && _coreJsData_default.keys.IE_PROTO || "");
	return uid ? "Symbol(src)_1." + uid : "";
}();
function isMasked(func) {
	return !!maskSrcKey && maskSrcKey in func;
}
var _isMasked_default = isMasked;
var funcToString$1 = Function.prototype.toString;
function toSource(func) {
	if (func != null) {
		try {
			return funcToString$1.call(func);
		} catch (e) {}
		try {
			return func + "";
		} catch (e) {}
	}
	return "";
}
var _toSource_default = toSource;
var reRegExpChar = /[\\^$.*+?()[\]{}|]/g;
var reIsHostCtor = /^\[object .+?Constructor\]$/;
var funcProto = Function.prototype, objectProto$2 = Object.prototype;
var funcToString = funcProto.toString;
var hasOwnProperty$8 = objectProto$2.hasOwnProperty;
var reIsNative = RegExp("^" + funcToString.call(hasOwnProperty$8).replace(reRegExpChar, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$");
function baseIsNative(value) {
	if (!isObject_default(value) || _isMasked_default(value)) return false;
	return (isFunction_default(value) ? reIsNative : reIsHostCtor).test(_toSource_default(value));
}
var _baseIsNative_default = baseIsNative;
function getValue(object, key) {
	return object == null ? void 0 : object[key];
}
var _getValue_default = getValue;
function getNative(object, key) {
	var value = _getValue_default(object, key);
	return _baseIsNative_default(value) ? value : void 0;
}
var _getNative_default = getNative;
var WeakMap = _getNative_default(_root_default, "WeakMap");
var _WeakMap_default = WeakMap;
var MAX_SAFE_INTEGER$1 = 9007199254740991;
var reIsUint = /^(?:0|[1-9]\d*)$/;
function isIndex(value, length) {
	var type = typeof value;
	length = length == null ? MAX_SAFE_INTEGER$1 : length;
	return !!length && (type == "number" || type != "symbol" && reIsUint.test(value)) && value > -1 && value % 1 == 0 && value < length;
}
var _isIndex_default = isIndex;
function eq(value, other) {
	return value === other || value !== value && other !== other;
}
var eq_default = eq;
var MAX_SAFE_INTEGER = 9007199254740991;
function isLength(value) {
	return typeof value == "number" && value > -1 && value % 1 == 0 && value <= MAX_SAFE_INTEGER;
}
var isLength_default = isLength;
function isArrayLike(value) {
	return value != null && isLength_default(value.length) && !isFunction_default(value);
}
var isArrayLike_default = isArrayLike;
var objectProto$1 = Object.prototype;
function isPrototype(value) {
	var Ctor = value && value.constructor;
	return value === (typeof Ctor == "function" && Ctor.prototype || objectProto$1);
}
var _isPrototype_default = isPrototype;
function baseTimes(n, iteratee) {
	var index = -1, result = Array(n);
	while (++index < n) result[index] = iteratee(index);
	return result;
}
var _baseTimes_default = baseTimes;
var argsTag$2 = "[object Arguments]";
function baseIsArguments(value) {
	return isObjectLike_default(value) && _baseGetTag_default(value) == argsTag$2;
}
var _baseIsArguments_default = baseIsArguments;
var objectProto = Object.prototype;
var hasOwnProperty$7 = objectProto.hasOwnProperty;
var propertyIsEnumerable$1 = objectProto.propertyIsEnumerable;
var isArguments = _baseIsArguments_default(function() {
	return arguments;
}()) ? _baseIsArguments_default : function(value) {
	return isObjectLike_default(value) && hasOwnProperty$7.call(value, "callee") && !propertyIsEnumerable$1.call(value, "callee");
};
var isArguments_default = isArguments;
function stubFalse() {
	return false;
}
var stubFalse_default = stubFalse;
var freeExports$1 = typeof exports == "object" && exports && !exports.nodeType && exports;
var freeModule$1 = freeExports$1 && typeof module == "object" && module && !module.nodeType && module;
var Buffer$1 = freeModule$1 && freeModule$1.exports === freeExports$1 ? _root_default.Buffer : void 0;
var isBuffer = (Buffer$1 ? Buffer$1.isBuffer : void 0) || stubFalse_default;
var isBuffer_default = isBuffer;
var argsTag$1 = "[object Arguments]", arrayTag$1 = "[object Array]", boolTag$1 = "[object Boolean]", dateTag$1 = "[object Date]", errorTag$1 = "[object Error]", funcTag = "[object Function]", mapTag$3 = "[object Map]", numberTag$1 = "[object Number]", objectTag$2 = "[object Object]", regexpTag$1 = "[object RegExp]", setTag$3 = "[object Set]", stringTag$1 = "[object String]", weakMapTag$1 = "[object WeakMap]";
var arrayBufferTag$1 = "[object ArrayBuffer]", dataViewTag$2 = "[object DataView]", float32Tag = "[object Float32Array]", float64Tag = "[object Float64Array]", int8Tag = "[object Int8Array]", int16Tag = "[object Int16Array]", int32Tag = "[object Int32Array]", uint8Tag = "[object Uint8Array]", uint8ClampedTag = "[object Uint8ClampedArray]", uint16Tag = "[object Uint16Array]", uint32Tag = "[object Uint32Array]";
var typedArrayTags = {};
typedArrayTags[float32Tag] = typedArrayTags[float64Tag] = typedArrayTags[int8Tag] = typedArrayTags[int16Tag] = typedArrayTags[int32Tag] = typedArrayTags[uint8Tag] = typedArrayTags[uint8ClampedTag] = typedArrayTags[uint16Tag] = typedArrayTags[uint32Tag] = true;
typedArrayTags[argsTag$1] = typedArrayTags[arrayTag$1] = typedArrayTags[arrayBufferTag$1] = typedArrayTags[boolTag$1] = typedArrayTags[dataViewTag$2] = typedArrayTags[dateTag$1] = typedArrayTags[errorTag$1] = typedArrayTags[funcTag] = typedArrayTags[mapTag$3] = typedArrayTags[numberTag$1] = typedArrayTags[objectTag$2] = typedArrayTags[regexpTag$1] = typedArrayTags[setTag$3] = typedArrayTags[stringTag$1] = typedArrayTags[weakMapTag$1] = false;
function baseIsTypedArray(value) {
	return isObjectLike_default(value) && isLength_default(value.length) && !!typedArrayTags[_baseGetTag_default(value)];
}
var _baseIsTypedArray_default = baseIsTypedArray;
function baseUnary(func) {
	return function(value) {
		return func(value);
	};
}
var _baseUnary_default = baseUnary;
var freeExports = typeof exports == "object" && exports && !exports.nodeType && exports;
var freeModule = freeExports && typeof module == "object" && module && !module.nodeType && module;
var freeProcess = freeModule && freeModule.exports === freeExports && _freeGlobal_default.process;
var nodeUtil = function() {
	try {
		var types = freeModule && freeModule.require && freeModule.require("util").types;
		if (types) return types;
		return freeProcess && freeProcess.binding && freeProcess.binding("util");
	} catch (e) {}
}();
var _nodeUtil_default = nodeUtil;
var nodeIsTypedArray = _nodeUtil_default && _nodeUtil_default.isTypedArray;
var isTypedArray = nodeIsTypedArray ? _baseUnary_default(nodeIsTypedArray) : _baseIsTypedArray_default;
var isTypedArray_default = isTypedArray;
var hasOwnProperty$6 = Object.prototype.hasOwnProperty;
function arrayLikeKeys(value, inherited) {
	var isArr = isArray_default(value), isArg = !isArr && isArguments_default(value), isBuff = !isArr && !isArg && isBuffer_default(value), isType = !isArr && !isArg && !isBuff && isTypedArray_default(value), skipIndexes = isArr || isArg || isBuff || isType, result = skipIndexes ? _baseTimes_default(value.length, String) : [], length = result.length;
	for (var key in value) if ((inherited || hasOwnProperty$6.call(value, key)) && !(skipIndexes && (key == "length" || isBuff && (key == "offset" || key == "parent") || isType && (key == "buffer" || key == "byteLength" || key == "byteOffset") || _isIndex_default(key, length)))) result.push(key);
	return result;
}
var _arrayLikeKeys_default = arrayLikeKeys;
function overArg(func, transform) {
	return function(arg) {
		return func(transform(arg));
	};
}
var _overArg_default = overArg;
var nativeKeys = _overArg_default(Object.keys, Object);
var _nativeKeys_default = nativeKeys;
var hasOwnProperty$5 = Object.prototype.hasOwnProperty;
function baseKeys(object) {
	if (!_isPrototype_default(object)) return _nativeKeys_default(object);
	var result = [];
	for (var key in Object(object)) if (hasOwnProperty$5.call(object, key) && key != "constructor") result.push(key);
	return result;
}
var _baseKeys_default = baseKeys;
function keys(object) {
	return isArrayLike_default(object) ? _arrayLikeKeys_default(object) : _baseKeys_default(object);
}
var keys_default = keys;
var reIsDeepProp = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/, reIsPlainProp = /^\w*$/;
function isKey(value, object) {
	if (isArray_default(value)) return false;
	var type = typeof value;
	if (type == "number" || type == "symbol" || type == "boolean" || value == null || isSymbol_default(value)) return true;
	return reIsPlainProp.test(value) || !reIsDeepProp.test(value) || object != null && value in Object(object);
}
var _isKey_default = isKey;
var nativeCreate = _getNative_default(Object, "create");
var _nativeCreate_default = nativeCreate;
function hashClear() {
	this.__data__ = _nativeCreate_default ? _nativeCreate_default(null) : {};
	this.size = 0;
}
var _hashClear_default = hashClear;
function hashDelete(key) {
	var result = this.has(key) && delete this.__data__[key];
	this.size -= result ? 1 : 0;
	return result;
}
var _hashDelete_default = hashDelete;
var HASH_UNDEFINED$2 = "__lodash_hash_undefined__";
var hasOwnProperty$4 = Object.prototype.hasOwnProperty;
function hashGet(key) {
	var data = this.__data__;
	if (_nativeCreate_default) {
		var result = data[key];
		return result === HASH_UNDEFINED$2 ? void 0 : result;
	}
	return hasOwnProperty$4.call(data, key) ? data[key] : void 0;
}
var _hashGet_default = hashGet;
var hasOwnProperty$3 = Object.prototype.hasOwnProperty;
function hashHas(key) {
	var data = this.__data__;
	return _nativeCreate_default ? data[key] !== void 0 : hasOwnProperty$3.call(data, key);
}
var _hashHas_default = hashHas;
var HASH_UNDEFINED$1 = "__lodash_hash_undefined__";
function hashSet(key, value) {
	var data = this.__data__;
	this.size += this.has(key) ? 0 : 1;
	data[key] = _nativeCreate_default && value === void 0 ? HASH_UNDEFINED$1 : value;
	return this;
}
var _hashSet_default = hashSet;
function Hash(entries) {
	var index = -1, length = entries == null ? 0 : entries.length;
	this.clear();
	while (++index < length) {
		var entry = entries[index];
		this.set(entry[0], entry[1]);
	}
}
Hash.prototype.clear = _hashClear_default;
Hash.prototype["delete"] = _hashDelete_default;
Hash.prototype.get = _hashGet_default;
Hash.prototype.has = _hashHas_default;
Hash.prototype.set = _hashSet_default;
var _Hash_default = Hash;
function listCacheClear() {
	this.__data__ = [];
	this.size = 0;
}
var _listCacheClear_default = listCacheClear;
function assocIndexOf(array, key) {
	var length = array.length;
	while (length--) if (eq_default(array[length][0], key)) return length;
	return -1;
}
var _assocIndexOf_default = assocIndexOf;
var splice = Array.prototype.splice;
function listCacheDelete(key) {
	var data = this.__data__, index = _assocIndexOf_default(data, key);
	if (index < 0) return false;
	if (index == data.length - 1) data.pop();
	else splice.call(data, index, 1);
	--this.size;
	return true;
}
var _listCacheDelete_default = listCacheDelete;
function listCacheGet(key) {
	var data = this.__data__, index = _assocIndexOf_default(data, key);
	return index < 0 ? void 0 : data[index][1];
}
var _listCacheGet_default = listCacheGet;
function listCacheHas(key) {
	return _assocIndexOf_default(this.__data__, key) > -1;
}
var _listCacheHas_default = listCacheHas;
function listCacheSet(key, value) {
	var data = this.__data__, index = _assocIndexOf_default(data, key);
	if (index < 0) {
		++this.size;
		data.push([key, value]);
	} else data[index][1] = value;
	return this;
}
var _listCacheSet_default = listCacheSet;
function ListCache(entries) {
	var index = -1, length = entries == null ? 0 : entries.length;
	this.clear();
	while (++index < length) {
		var entry = entries[index];
		this.set(entry[0], entry[1]);
	}
}
ListCache.prototype.clear = _listCacheClear_default;
ListCache.prototype["delete"] = _listCacheDelete_default;
ListCache.prototype.get = _listCacheGet_default;
ListCache.prototype.has = _listCacheHas_default;
ListCache.prototype.set = _listCacheSet_default;
var _ListCache_default = ListCache;
var Map = _getNative_default(_root_default, "Map");
var _Map_default = Map;
function mapCacheClear() {
	this.size = 0;
	this.__data__ = {
		"hash": new _Hash_default(),
		"map": new (_Map_default || _ListCache_default)(),
		"string": new _Hash_default()
	};
}
var _mapCacheClear_default = mapCacheClear;
function isKeyable(value) {
	var type = typeof value;
	return type == "string" || type == "number" || type == "symbol" || type == "boolean" ? value !== "__proto__" : value === null;
}
var _isKeyable_default = isKeyable;
function getMapData(map, key) {
	var data = map.__data__;
	return _isKeyable_default(key) ? data[typeof key == "string" ? "string" : "hash"] : data.map;
}
var _getMapData_default = getMapData;
function mapCacheDelete(key) {
	var result = _getMapData_default(this, key)["delete"](key);
	this.size -= result ? 1 : 0;
	return result;
}
var _mapCacheDelete_default = mapCacheDelete;
function mapCacheGet(key) {
	return _getMapData_default(this, key).get(key);
}
var _mapCacheGet_default = mapCacheGet;
function mapCacheHas(key) {
	return _getMapData_default(this, key).has(key);
}
var _mapCacheHas_default = mapCacheHas;
function mapCacheSet(key, value) {
	var data = _getMapData_default(this, key), size = data.size;
	data.set(key, value);
	this.size += data.size == size ? 0 : 1;
	return this;
}
var _mapCacheSet_default = mapCacheSet;
function MapCache(entries) {
	var index = -1, length = entries == null ? 0 : entries.length;
	this.clear();
	while (++index < length) {
		var entry = entries[index];
		this.set(entry[0], entry[1]);
	}
}
MapCache.prototype.clear = _mapCacheClear_default;
MapCache.prototype["delete"] = _mapCacheDelete_default;
MapCache.prototype.get = _mapCacheGet_default;
MapCache.prototype.has = _mapCacheHas_default;
MapCache.prototype.set = _mapCacheSet_default;
var _MapCache_default = MapCache;
var FUNC_ERROR_TEXT = "Expected a function";
function memoize(func, resolver) {
	if (typeof func != "function" || resolver != null && typeof resolver != "function") throw new TypeError(FUNC_ERROR_TEXT);
	var memoized = function() {
		var args = arguments, key = resolver ? resolver.apply(this, args) : args[0], cache = memoized.cache;
		if (cache.has(key)) return cache.get(key);
		var result = func.apply(this, args);
		memoized.cache = cache.set(key, result) || cache;
		return result;
	};
	memoized.cache = new (memoize.Cache || _MapCache_default)();
	return memoized;
}
memoize.Cache = _MapCache_default;
var memoize_default = memoize;
var MAX_MEMOIZE_SIZE = 500;
function memoizeCapped(func) {
	var result = memoize_default(func, function(key) {
		if (cache.size === MAX_MEMOIZE_SIZE) cache.clear();
		return key;
	});
	var cache = result.cache;
	return result;
}
var _memoizeCapped_default = memoizeCapped;
var rePropName = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g;
var reEscapeChar = /\\(\\)?/g;
var stringToPath = _memoizeCapped_default(function(string) {
	var result = [];
	if (string.charCodeAt(0) === 46) result.push("");
	string.replace(rePropName, function(match, number, quote, subString) {
		result.push(quote ? subString.replace(reEscapeChar, "$1") : number || match);
	});
	return result;
});
var _stringToPath_default = stringToPath;
function toString(value) {
	return value == null ? "" : _baseToString_default(value);
}
var toString_default = toString;
function castPath(value, object) {
	if (isArray_default(value)) return value;
	return _isKey_default(value, object) ? [value] : _stringToPath_default(toString_default(value));
}
var _castPath_default = castPath;
var INFINITY = Infinity;
function toKey(value) {
	if (typeof value == "string" || isSymbol_default(value)) return value;
	var result = value + "";
	return result == "0" && 1 / value == -INFINITY ? "-0" : result;
}
var _toKey_default = toKey;
function baseGet(object, path$6) {
	path$6 = _castPath_default(path$6, object);
	var index = 0, length = path$6.length;
	while (object != null && index < length) object = object[_toKey_default(path$6[index++])];
	return index && index == length ? object : void 0;
}
var _baseGet_default = baseGet;
function get(object, path$6, defaultValue) {
	var result = object == null ? void 0 : _baseGet_default(object, path$6);
	return result === void 0 ? defaultValue : result;
}
var get_default = get;
function arrayPush(array, values) {
	var index = -1, length = values.length, offset = array.length;
	while (++index < length) array[offset + index] = values[index];
	return array;
}
var _arrayPush_default = arrayPush;
function arrayReduce(array, iteratee, accumulator, initAccum) {
	var index = -1, length = array == null ? 0 : array.length;
	if (initAccum && length) accumulator = array[++index];
	while (++index < length) accumulator = iteratee(accumulator, array[index], index, array);
	return accumulator;
}
var _arrayReduce_default = arrayReduce;
function stackClear() {
	this.__data__ = new _ListCache_default();
	this.size = 0;
}
var _stackClear_default = stackClear;
function stackDelete(key) {
	var data = this.__data__, result = data["delete"](key);
	this.size = data.size;
	return result;
}
var _stackDelete_default = stackDelete;
function stackGet(key) {
	return this.__data__.get(key);
}
var _stackGet_default = stackGet;
function stackHas(key) {
	return this.__data__.has(key);
}
var _stackHas_default = stackHas;
var LARGE_ARRAY_SIZE = 200;
function stackSet(key, value) {
	var data = this.__data__;
	if (data instanceof _ListCache_default) {
		var pairs = data.__data__;
		if (!_Map_default || pairs.length < LARGE_ARRAY_SIZE - 1) {
			pairs.push([key, value]);
			this.size = ++data.size;
			return this;
		}
		data = this.__data__ = new _MapCache_default(pairs);
	}
	data.set(key, value);
	this.size = data.size;
	return this;
}
var _stackSet_default = stackSet;
function Stack(entries) {
	this.size = (this.__data__ = new _ListCache_default(entries)).size;
}
Stack.prototype.clear = _stackClear_default;
Stack.prototype["delete"] = _stackDelete_default;
Stack.prototype.get = _stackGet_default;
Stack.prototype.has = _stackHas_default;
Stack.prototype.set = _stackSet_default;
var _Stack_default = Stack;
function arrayFilter(array, predicate) {
	var index = -1, length = array == null ? 0 : array.length, resIndex = 0, result = [];
	while (++index < length) {
		var value = array[index];
		if (predicate(value, index, array)) result[resIndex++] = value;
	}
	return result;
}
var _arrayFilter_default = arrayFilter;
function stubArray() {
	return [];
}
var stubArray_default = stubArray;
var propertyIsEnumerable = Object.prototype.propertyIsEnumerable;
var nativeGetSymbols = Object.getOwnPropertySymbols;
var getSymbols = !nativeGetSymbols ? stubArray_default : function(object) {
	if (object == null) return [];
	object = Object(object);
	return _arrayFilter_default(nativeGetSymbols(object), function(symbol) {
		return propertyIsEnumerable.call(object, symbol);
	});
};
var _getSymbols_default = getSymbols;
function baseGetAllKeys(object, keysFunc, symbolsFunc) {
	var result = keysFunc(object);
	return isArray_default(object) ? result : _arrayPush_default(result, symbolsFunc(object));
}
var _baseGetAllKeys_default = baseGetAllKeys;
function getAllKeys(object) {
	return _baseGetAllKeys_default(object, keys_default, _getSymbols_default);
}
var _getAllKeys_default = getAllKeys;
var DataView = _getNative_default(_root_default, "DataView");
var _DataView_default = DataView;
var Promise$1 = _getNative_default(_root_default, "Promise");
var _Promise_default = Promise$1;
var Set = _getNative_default(_root_default, "Set");
var _Set_default = Set;
var mapTag$2 = "[object Map]", objectTag$1 = "[object Object]", promiseTag = "[object Promise]", setTag$2 = "[object Set]", weakMapTag = "[object WeakMap]";
var dataViewTag$1 = "[object DataView]";
var dataViewCtorString = _toSource_default(_DataView_default), mapCtorString = _toSource_default(_Map_default), promiseCtorString = _toSource_default(_Promise_default), setCtorString = _toSource_default(_Set_default), weakMapCtorString = _toSource_default(_WeakMap_default);
var getTag = _baseGetTag_default;
if (_DataView_default && getTag(new _DataView_default(/* @__PURE__ */ new ArrayBuffer(1))) != dataViewTag$1 || _Map_default && getTag(new _Map_default()) != mapTag$2 || _Promise_default && getTag(_Promise_default.resolve()) != promiseTag || _Set_default && getTag(new _Set_default()) != setTag$2 || _WeakMap_default && getTag(new _WeakMap_default()) != weakMapTag) getTag = function(value) {
	var result = _baseGetTag_default(value), Ctor = result == objectTag$1 ? value.constructor : void 0, ctorString = Ctor ? _toSource_default(Ctor) : "";
	if (ctorString) switch (ctorString) {
		case dataViewCtorString: return dataViewTag$1;
		case mapCtorString: return mapTag$2;
		case promiseCtorString: return promiseTag;
		case setCtorString: return setTag$2;
		case weakMapCtorString: return weakMapTag;
	}
	return result;
};
var _getTag_default = getTag;
var Uint8Array$1 = _root_default.Uint8Array;
var _Uint8Array_default = Uint8Array$1;
var HASH_UNDEFINED = "__lodash_hash_undefined__";
function setCacheAdd(value) {
	this.__data__.set(value, HASH_UNDEFINED);
	return this;
}
var _setCacheAdd_default = setCacheAdd;
function setCacheHas(value) {
	return this.__data__.has(value);
}
var _setCacheHas_default = setCacheHas;
function SetCache(values) {
	var index = -1, length = values == null ? 0 : values.length;
	this.__data__ = new _MapCache_default();
	while (++index < length) this.add(values[index]);
}
SetCache.prototype.add = SetCache.prototype.push = _setCacheAdd_default;
SetCache.prototype.has = _setCacheHas_default;
var _SetCache_default = SetCache;
function arraySome(array, predicate) {
	var index = -1, length = array == null ? 0 : array.length;
	while (++index < length) if (predicate(array[index], index, array)) return true;
	return false;
}
var _arraySome_default = arraySome;
function cacheHas(cache, key) {
	return cache.has(key);
}
var _cacheHas_default = cacheHas;
var COMPARE_PARTIAL_FLAG$5 = 1, COMPARE_UNORDERED_FLAG$3 = 2;
function equalArrays(array, other, bitmask, customizer, equalFunc, stack) {
	var isPartial = bitmask & COMPARE_PARTIAL_FLAG$5, arrLength = array.length, othLength = other.length;
	if (arrLength != othLength && !(isPartial && othLength > arrLength)) return false;
	var arrStacked = stack.get(array);
	var othStacked = stack.get(other);
	if (arrStacked && othStacked) return arrStacked == other && othStacked == array;
	var index = -1, result = true, seen = bitmask & COMPARE_UNORDERED_FLAG$3 ? new _SetCache_default() : void 0;
	stack.set(array, other);
	stack.set(other, array);
	while (++index < arrLength) {
		var arrValue = array[index], othValue = other[index];
		if (customizer) var compared = isPartial ? customizer(othValue, arrValue, index, other, array, stack) : customizer(arrValue, othValue, index, array, other, stack);
		if (compared !== void 0) {
			if (compared) continue;
			result = false;
			break;
		}
		if (seen) {
			if (!_arraySome_default(other, function(othValue$1, othIndex) {
				if (!_cacheHas_default(seen, othIndex) && (arrValue === othValue$1 || equalFunc(arrValue, othValue$1, bitmask, customizer, stack))) return seen.push(othIndex);
			})) {
				result = false;
				break;
			}
		} else if (!(arrValue === othValue || equalFunc(arrValue, othValue, bitmask, customizer, stack))) {
			result = false;
			break;
		}
	}
	stack["delete"](array);
	stack["delete"](other);
	return result;
}
var _equalArrays_default = equalArrays;
function mapToArray(map) {
	var index = -1, result = Array(map.size);
	map.forEach(function(value, key) {
		result[++index] = [key, value];
	});
	return result;
}
var _mapToArray_default = mapToArray;
function setToArray(set) {
	var index = -1, result = Array(set.size);
	set.forEach(function(value) {
		result[++index] = value;
	});
	return result;
}
var _setToArray_default = setToArray;
var COMPARE_PARTIAL_FLAG$4 = 1, COMPARE_UNORDERED_FLAG$2 = 2;
var boolTag = "[object Boolean]", dateTag = "[object Date]", errorTag = "[object Error]", mapTag$1 = "[object Map]", numberTag = "[object Number]", regexpTag = "[object RegExp]", setTag$1 = "[object Set]", stringTag = "[object String]", symbolTag = "[object Symbol]";
var arrayBufferTag = "[object ArrayBuffer]", dataViewTag = "[object DataView]";
var symbolProto = _Symbol_default ? _Symbol_default.prototype : void 0, symbolValueOf = symbolProto ? symbolProto.valueOf : void 0;
function equalByTag(object, other, tag, bitmask, customizer, equalFunc, stack) {
	switch (tag) {
		case dataViewTag:
			if (object.byteLength != other.byteLength || object.byteOffset != other.byteOffset) return false;
			object = object.buffer;
			other = other.buffer;
		case arrayBufferTag:
			if (object.byteLength != other.byteLength || !equalFunc(new _Uint8Array_default(object), new _Uint8Array_default(other))) return false;
			return true;
		case boolTag:
		case dateTag:
		case numberTag: return eq_default(+object, +other);
		case errorTag: return object.name == other.name && object.message == other.message;
		case regexpTag:
		case stringTag: return object == other + "";
		case mapTag$1: var convert = _mapToArray_default;
		case setTag$1:
			var isPartial = bitmask & COMPARE_PARTIAL_FLAG$4;
			convert || (convert = _setToArray_default);
			if (object.size != other.size && !isPartial) return false;
			var stacked = stack.get(object);
			if (stacked) return stacked == other;
			bitmask |= COMPARE_UNORDERED_FLAG$2;
			stack.set(object, other);
			var result = _equalArrays_default(convert(object), convert(other), bitmask, customizer, equalFunc, stack);
			stack["delete"](object);
			return result;
		case symbolTag: if (symbolValueOf) return symbolValueOf.call(object) == symbolValueOf.call(other);
	}
	return false;
}
var _equalByTag_default = equalByTag;
var COMPARE_PARTIAL_FLAG$3 = 1;
var hasOwnProperty$2 = Object.prototype.hasOwnProperty;
function equalObjects(object, other, bitmask, customizer, equalFunc, stack) {
	var isPartial = bitmask & COMPARE_PARTIAL_FLAG$3, objProps = _getAllKeys_default(object), objLength = objProps.length;
	if (objLength != _getAllKeys_default(other).length && !isPartial) return false;
	var index = objLength;
	while (index--) {
		var key = objProps[index];
		if (!(isPartial ? key in other : hasOwnProperty$2.call(other, key))) return false;
	}
	var objStacked = stack.get(object);
	var othStacked = stack.get(other);
	if (objStacked && othStacked) return objStacked == other && othStacked == object;
	var result = true;
	stack.set(object, other);
	stack.set(other, object);
	var skipCtor = isPartial;
	while (++index < objLength) {
		key = objProps[index];
		var objValue = object[key], othValue = other[key];
		if (customizer) var compared = isPartial ? customizer(othValue, objValue, key, other, object, stack) : customizer(objValue, othValue, key, object, other, stack);
		if (!(compared === void 0 ? objValue === othValue || equalFunc(objValue, othValue, bitmask, customizer, stack) : compared)) {
			result = false;
			break;
		}
		skipCtor || (skipCtor = key == "constructor");
	}
	if (result && !skipCtor) {
		var objCtor = object.constructor, othCtor = other.constructor;
		if (objCtor != othCtor && "constructor" in object && "constructor" in other && !(typeof objCtor == "function" && objCtor instanceof objCtor && typeof othCtor == "function" && othCtor instanceof othCtor)) result = false;
	}
	stack["delete"](object);
	stack["delete"](other);
	return result;
}
var _equalObjects_default = equalObjects;
var COMPARE_PARTIAL_FLAG$2 = 1;
var argsTag = "[object Arguments]", arrayTag = "[object Array]", objectTag = "[object Object]";
var hasOwnProperty$1 = Object.prototype.hasOwnProperty;
function baseIsEqualDeep(object, other, bitmask, customizer, equalFunc, stack) {
	var objIsArr = isArray_default(object), othIsArr = isArray_default(other), objTag = objIsArr ? arrayTag : _getTag_default(object), othTag = othIsArr ? arrayTag : _getTag_default(other);
	objTag = objTag == argsTag ? objectTag : objTag;
	othTag = othTag == argsTag ? objectTag : othTag;
	var objIsObj = objTag == objectTag, othIsObj = othTag == objectTag, isSameTag = objTag == othTag;
	if (isSameTag && isBuffer_default(object)) {
		if (!isBuffer_default(other)) return false;
		objIsArr = true;
		objIsObj = false;
	}
	if (isSameTag && !objIsObj) {
		stack || (stack = new _Stack_default());
		return objIsArr || isTypedArray_default(object) ? _equalArrays_default(object, other, bitmask, customizer, equalFunc, stack) : _equalByTag_default(object, other, objTag, bitmask, customizer, equalFunc, stack);
	}
	if (!(bitmask & COMPARE_PARTIAL_FLAG$2)) {
		var objIsWrapped = objIsObj && hasOwnProperty$1.call(object, "__wrapped__"), othIsWrapped = othIsObj && hasOwnProperty$1.call(other, "__wrapped__");
		if (objIsWrapped || othIsWrapped) {
			var objUnwrapped = objIsWrapped ? object.value() : object, othUnwrapped = othIsWrapped ? other.value() : other;
			stack || (stack = new _Stack_default());
			return equalFunc(objUnwrapped, othUnwrapped, bitmask, customizer, stack);
		}
	}
	if (!isSameTag) return false;
	stack || (stack = new _Stack_default());
	return _equalObjects_default(object, other, bitmask, customizer, equalFunc, stack);
}
var _baseIsEqualDeep_default = baseIsEqualDeep;
function baseIsEqual(value, other, bitmask, customizer, stack) {
	if (value === other) return true;
	if (value == null || other == null || !isObjectLike_default(value) && !isObjectLike_default(other)) return value !== value && other !== other;
	return _baseIsEqualDeep_default(value, other, bitmask, customizer, baseIsEqual, stack);
}
var _baseIsEqual_default = baseIsEqual;
var COMPARE_PARTIAL_FLAG$1 = 1, COMPARE_UNORDERED_FLAG$1 = 2;
function baseIsMatch(object, source, matchData, customizer) {
	var index = matchData.length, length = index, noCustomizer = !customizer;
	if (object == null) return !length;
	object = Object(object);
	while (index--) {
		var data = matchData[index];
		if (noCustomizer && data[2] ? data[1] !== object[data[0]] : !(data[0] in object)) return false;
	}
	while (++index < length) {
		data = matchData[index];
		var key = data[0], objValue = object[key], srcValue = data[1];
		if (noCustomizer && data[2]) {
			if (objValue === void 0 && !(key in object)) return false;
		} else {
			var stack = new _Stack_default();
			if (customizer) var result = customizer(objValue, srcValue, key, object, source, stack);
			if (!(result === void 0 ? _baseIsEqual_default(srcValue, objValue, COMPARE_PARTIAL_FLAG$1 | COMPARE_UNORDERED_FLAG$1, customizer, stack) : result)) return false;
		}
	}
	return true;
}
var _baseIsMatch_default = baseIsMatch;
function isStrictComparable(value) {
	return value === value && !isObject_default(value);
}
var _isStrictComparable_default = isStrictComparable;
function getMatchData(object) {
	var result = keys_default(object), length = result.length;
	while (length--) {
		var key = result[length], value = object[key];
		result[length] = [
			key,
			value,
			_isStrictComparable_default(value)
		];
	}
	return result;
}
var _getMatchData_default = getMatchData;
function matchesStrictComparable(key, srcValue) {
	return function(object) {
		if (object == null) return false;
		return object[key] === srcValue && (srcValue !== void 0 || key in Object(object));
	};
}
var _matchesStrictComparable_default = matchesStrictComparable;
function baseMatches(source) {
	var matchData = _getMatchData_default(source);
	if (matchData.length == 1 && matchData[0][2]) return _matchesStrictComparable_default(matchData[0][0], matchData[0][1]);
	return function(object) {
		return object === source || _baseIsMatch_default(object, source, matchData);
	};
}
var _baseMatches_default = baseMatches;
function baseHasIn(object, key) {
	return object != null && key in Object(object);
}
var _baseHasIn_default = baseHasIn;
function hasPath(object, path$6, hasFunc) {
	path$6 = _castPath_default(path$6, object);
	var index = -1, length = path$6.length, result = false;
	while (++index < length) {
		var key = _toKey_default(path$6[index]);
		if (!(result = object != null && hasFunc(object, key))) break;
		object = object[key];
	}
	if (result || ++index != length) return result;
	length = object == null ? 0 : object.length;
	return !!length && isLength_default(length) && _isIndex_default(key, length) && (isArray_default(object) || isArguments_default(object));
}
var _hasPath_default = hasPath;
function hasIn(object, path$6) {
	return object != null && _hasPath_default(object, path$6, _baseHasIn_default);
}
var hasIn_default = hasIn;
var COMPARE_PARTIAL_FLAG = 1, COMPARE_UNORDERED_FLAG = 2;
function baseMatchesProperty(path$6, srcValue) {
	if (_isKey_default(path$6) && _isStrictComparable_default(srcValue)) return _matchesStrictComparable_default(_toKey_default(path$6), srcValue);
	return function(object) {
		var objValue = get_default(object, path$6);
		return objValue === void 0 && objValue === srcValue ? hasIn_default(object, path$6) : _baseIsEqual_default(srcValue, objValue, COMPARE_PARTIAL_FLAG | COMPARE_UNORDERED_FLAG);
	};
}
var _baseMatchesProperty_default = baseMatchesProperty;
function baseProperty(key) {
	return function(object) {
		return object == null ? void 0 : object[key];
	};
}
var _baseProperty_default = baseProperty;
function basePropertyDeep(path$6) {
	return function(object) {
		return _baseGet_default(object, path$6);
	};
}
var _basePropertyDeep_default = basePropertyDeep;
function property(path$6) {
	return _isKey_default(path$6) ? _baseProperty_default(_toKey_default(path$6)) : _basePropertyDeep_default(path$6);
}
var property_default = property;
function baseIteratee(value) {
	if (typeof value == "function") return value;
	if (value == null) return identity_default;
	if (typeof value == "object") return isArray_default(value) ? _baseMatchesProperty_default(value[0], value[1]) : _baseMatches_default(value);
	return property_default(value);
}
var _baseIteratee_default = baseIteratee;
function createBaseFor(fromRight) {
	return function(object, iteratee, keysFunc) {
		var index = -1, iterable = Object(object), props = keysFunc(object), length = props.length;
		while (length--) {
			var key = props[fromRight ? length : ++index];
			if (iteratee(iterable[key], key, iterable) === false) break;
		}
		return object;
	};
}
var _createBaseFor_default = createBaseFor;
var baseFor = _createBaseFor_default();
var _baseFor_default = baseFor;
function baseForOwn(object, iteratee) {
	return object && _baseFor_default(object, iteratee, keys_default);
}
var _baseForOwn_default = baseForOwn;
function createBaseEach(eachFunc, fromRight) {
	return function(collection, iteratee) {
		if (collection == null) return collection;
		if (!isArrayLike_default(collection)) return eachFunc(collection, iteratee);
		var length = collection.length, index = fromRight ? length : -1, iterable = Object(collection);
		while (fromRight ? index-- : ++index < length) if (iteratee(iterable[index], index, iterable) === false) break;
		return collection;
	};
}
var _createBaseEach_default = createBaseEach;
var baseEach = _createBaseEach_default(_baseForOwn_default);
var _baseEach_default = baseEach;
var mapTag = "[object Map]", setTag = "[object Set]";
var hasOwnProperty = Object.prototype.hasOwnProperty;
function isEmpty(value) {
	if (value == null) return true;
	if (isArrayLike_default(value) && (isArray_default(value) || typeof value == "string" || typeof value.splice == "function" || isBuffer_default(value) || isTypedArray_default(value) || isArguments_default(value))) return !value.length;
	var tag = _getTag_default(value);
	if (tag == mapTag || tag == setTag) return !value.size;
	if (_isPrototype_default(value)) return !_baseKeys_default(value).length;
	for (var key in value) if (hasOwnProperty.call(value, key)) return false;
	return true;
}
var isEmpty_default = isEmpty;
function baseReduce(collection, iteratee, accumulator, initAccum, eachFunc) {
	eachFunc(collection, function(value, index, collection$1) {
		accumulator = initAccum ? (initAccum = false, value) : iteratee(accumulator, value, index, collection$1);
	});
	return accumulator;
}
var _baseReduce_default = baseReduce;
function reduce(collection, iteratee, accumulator) {
	var func = isArray_default(collection) ? _arrayReduce_default : _baseReduce_default, initAccum = arguments.length < 3;
	return func(collection, _baseIteratee_default(iteratee, 4), accumulator, initAccum, _baseEach_default);
}
var reduce_default = reduce;
function resolvePaths(paths) {
	if (!Array.isArray(paths)) return path.default.resolve(process.cwd(), paths);
	return paths.map((_path) => path.default.resolve(process.cwd(), _path));
}
function validateInputFilePaths(paths, validatePlugin) {
	return paths.every((_path) => validatePlugin(_path));
}
function getFormatOpt(opts) {
	const formatOpts = Object.keys(opts).filter((opt) => {
		return [
			"postgres",
			"mysql",
			"mssql",
			"postgresLegacy",
			"mysqlLegacy",
			"mssqlLegacy",
			"oracle",
			"snowflake"
		].includes(opt);
	});
	let format$1 = "postgres";
	let cnt = 0;
	formatOpts.forEach((opt) => {
		if (opts[opt]) {
			cnt += 1;
			if (cnt > 1) throw new Error("Too many format options");
			format$1 = opt;
		}
	});
	return format$1;
}
function getConnectionOpt(args) {
	const supportedDatabases = [
		"postgres",
		"mysql",
		"mssql",
		"snowflake",
		"bigquery",
		"oracle"
	];
	return reduce_default(args, (connectionOpt, arg) => {
		if (supportedDatabases.includes(arg)) connectionOpt.databaseType = arg;
		if (/^.*[:;]/.test(arg)) connectionOpt.connection = arg;
		if (/^[a-zA-Z]:[\\/](?:[^<>:"/\\|?*\n\r]+[\\/])*[^<>:"/\\|?*\n\r]*$/.test(arg) || /^(\/|\.\/|~\/|\.\.\/)([^<>:"|?*\n\r]*\/?)*[^<>:"|?*\n\r]*$/.test(arg)) connectionOpt.connection = arg;
		return connectionOpt;
	}, {
		connection: args[0],
		databaseType: "unknown"
	});
}
function generate(inputPaths, transform, outputPlugin) {
	inputPaths.forEach((_path) => {
		const source = fs.default.readFileSync(_path, "utf-8");
		try {
			const content = transform(source);
			outputPlugin.write(content);
		} catch (e) {
			if (e instanceof __dbml_core.CompilerError) throw e.map((diag) => ({
				...diag,
				message: diag.message,
				filepath: path.default.basename(_path),
				stack: diag.stack
			}));
			throw e;
		}
	});
}
async function connectionHandler(program$1) {
	try {
		const { connection, databaseType } = getConnectionOpt(program$1.args);
		const opts = program$1.opts();
		const schemaJson = await __dbml_connector.connector.fetchSchemaJson(connection, databaseType);
		if (!opts.outFile && !opts.outDir) {
			const res = __dbml_core.importer.generateDbml(schemaJson);
			outputConsolePlugin_default.write(res);
		} else if (opts.outFile) {
			const res = __dbml_core.importer.generateDbml(schemaJson);
			new outputFilePlugin_default(resolvePaths(opts.outFile)).write(res);
			console.log(`  ${chalk.default.green(figures.default.main.tick)} Generated DBML file from database's connection: ${path.default.basename(opts.outFile)}`);
		}
	} catch (error) {
		consoleLogger.error(error);
		fileLogger.error(error);
	}
}
var NodeProjectLayout = class {
	constructor(entryPoints) {
		this.entryPoints = [...entryPoints];
	}
	getSource(filePath) {
		return safeReadFileSync(filePath.absolute, "utf-8");
	}
	exists(filePath) {
		return this.isFile(filePath) || this.isDirectory(filePath);
	}
	isFile(filePath) {
		return safeStatSync(filePath.absolute, { throwIfNoEntry: false })?.isFile() ?? false;
	}
	isDirectory(filePath) {
		return safeStatSync(filePath.absolute, { throwIfNoEntry: false })?.isDirectory() ?? false;
	}
	listDirectory(dirPath) {
		const basePath = dirPath?.absolute ?? "/";
		return safeReaddirSync(basePath)?.map((entry) => basePath + entry).sort().map(__dbml_parse.Filepath.from) ?? [];
	}
	getEntrypoints() {
		return this.entryPoints.filter((fp) => this.exists(fp));
	}
};
function logOnError(fn) {
	return (...args) => {
		try {
			return fn(...args);
		} catch (e) {
			fileLogger.error(e instanceof Error ? e : new Error(String(e)));
			return;
		}
	};
}
var safeReadFileSync = logOnError(node_fs.readFileSync);
var safeReaddirSync = logOnError(node_fs.readdirSync);
var safeStatSync = logOnError((path$6, opts) => (0, node_fs.statSync)(path$6, opts));
var config = {
	mysql: { name: "MySQL" },
	mysqlLegacy: { name: "MySQL" },
	postgres: { name: "PostgreSQL" },
	postgresLegacy: { name: "PostgreSQL" },
	mssql: { name: "SQL Server" },
	oracle: { name: "Oracle" },
	snowflake: { name: "Snowflake" }
};
var config_default = config;
function validateFilePlugin(_path) {
	if (fs.default.statSync(_path).isDirectory()) throw new Error("Expect input to be files");
}
async function exportHandler(program$1) {
	const inputPaths = resolvePaths(program$1.args);
	validateInputFilePaths(inputPaths, validateFilePlugin);
	const opts = program$1.opts();
	const format$1 = getFormatOpt(opts);
	const entrypoints = inputPaths.map((p) => new __dbml_parse.Filepath(p));
	const compiler = new __dbml_parse.Compiler(new NodeProjectLayout(entrypoints));
	const outputPlugin = opts.outFile ? new outputFilePlugin_default(resolvePaths(opts.outFile), [
		"-- SQL dump generated using DBML (dbml.dbdiagram.io)\n",
		`-- Database: ${config_default[format$1].name}\n`,
		`-- Generated at: ${(/* @__PURE__ */ new Date()).toISOString()}\n\n`
	].join("")) : outputConsolePlugin_default;
	let hasErrors = false;
	for (const filepath of entrypoints) {
		const result = compiler.interpretFile(filepath);
		const errors = result.getErrors();
		if (errors.length > 0) {
			hasErrors = true;
			const isMultifile = compiler.reachableFiles(filepath).length > 1;
			const msg = `\n    ${errors.map((e) => {
				const pos = e.nodeOrToken.startPos;
				const line = pos.line + 1;
				const col = pos.column + 1;
				return `${chalk.default.cyan(isMultifile ? e.filepath.relativeTo(process.cwd()) : e.filepath.basename)}${chalk.default.yellow(`(${line},${col})`)}: ${e.message}`;
			}).join("\n    ")}`;
			consoleLogger.error(msg);
			fileLogger.error(msg);
			continue;
		}
		const db = result.getValue();
		if (!db) continue;
		const sql = __dbml_core.ModelExporter.export(__dbml_core.Parser.parseJSONToDatabase(db), format$1);
		outputPlugin.write(sql);
	}
	if (!hasErrors && opts.outFile) console.log(`  ${chalk.default.green(figures.default.main.tick)} Generated SQL dump file (${config_default[format$1].name}): ${path.default.basename(opts.outFile)}`);
}
var DomainError = class extends Error {
	constructor(message, rootError = {}) {
		super(message);
		this.name = this.constructor.name;
		this.rootError = rootError;
		Error.captureStackTrace(this, this.constructor);
	}
};
var domainError_default = DomainError;
var SyntaxError = class extends domainError_default {
	constructor(fileName, rootError = {}) {
		let message = `You have a syntax error at "${fileName}"`;
		if (rootError.location) message += ` line ${rootError.location.start.line} column ${rootError.location.start.column}`;
		message += ".";
		if (!isEmpty_default(rootError)) message += ` ${rootError.message}`;
		super(message, rootError);
	}
};
var syntaxError_default = SyntaxError;
async function importHandler(program$1) {
	try {
		const inputPaths = resolvePaths(program$1.args);
		validateInputFilePaths(inputPaths, validateFilePlugin);
		const opts = program$1.opts();
		const format$1 = getFormatOpt(opts);
		if (!opts.outFile && !opts.outDir) generate(inputPaths, (sql) => __dbml_core.importer.import(sql, format$1), outputConsolePlugin_default);
		else if (opts.outFile) {
			generate(inputPaths, (sql) => __dbml_core.importer.import(sql, format$1), new outputFilePlugin_default(resolvePaths(opts.outFile)));
			console.log(`  ${chalk.default.green(figures.default.main.tick)} Generated DBML file from SQL file (${config_default[format$1].name}): ${path.default.basename(opts.outFile)}`);
		}
	} catch (error) {
		if (error instanceof __dbml_core.CompilerError) {
			const msg = `\n    ${error.diags.map((diag) => new syntaxError_default(diag.filepath ?? "", diag)).map(({ message }) => message).join("\n    ")}`;
			consoleLogger.error(msg);
			fileLogger.error(msg);
			return;
		}
		throw error;
	}
}
function showHelp(args) {
	if (args.length < 3) commander.default.help();
}
function dbml2sql(args) {
	commander.default.version(package_default.version);
	commander.default.usage("[options] <files...>").option("--mysql").option("--postgres").option("--mssql").option("--oracle").option("-o, --out-file <pathspec>", "compile all input files into a single files");
	showHelp(args);
	commander.default.parse(args);
	exportHandler(commander.default);
}
function sql2dbml(args) {
	commander.default.version(package_default.version);
	commander.default.usage("[options] <files...>").option("--mysql").option("--mysql-legacy").option("--postgres").option("--postgres-legacy").option("--mssql").option("--mssql-legacy").option("--snowflake").option("--oracle").option("-o, --out-file <pathspec>", "compile all input files into a single files");
	showHelp(args);
	commander.default.parse(args);
	importHandler(commander.default);
}
function db2dbml(args) {
	commander.default.version(package_default.version);
	commander.default.usage("<database-type> <connection-string> [options]").description(`Generate DBML directly from a database
    <database-type>     your database format (postgres, mysql, mssql, snowflake, bigquery)
    <connection-string> your database connection string:
                        - postgres: 'postgresql://user:password@localhost:5432/dbname?schemas=schema1,schema2,schema3'
                        - mysql: 'mysql://user:password@localhost:3306/dbname'
                        - mssql: 'Server=localhost,1433;Database=master;User Id=sa;Password=your_password;Encrypt=true;TrustServerCertificate=true;Schemas=schema1,schema2,schema3;'
                        - oracle: 'username/password@[//]host[:port][/service_name]'
                        - snowflake:
                          - password-based authentication: 'SERVER=<account_identifier>.<region>;UID=<your_username>;PWD=<your_password>;DATABASE=<your_database>;WAREHOUSE=<your_warehouse>;ROLE=<your_role>;SCHEMAS=schema1,schema2,schema3;'
                          - key pair authentication: 'SERVER=<account_identifier>.<region>;UID=<your_username>;AUTHENTICATOR=SNOWFLAKE_JWT;PRIVATE_KEY_PATH=<path_to_your_private_key.p8>;PASSPHRASE=<your_private_key_passphrase>;DATABASE=<your_database>;WAREHOUSE=<your_warehouse>;ROLE=<your_role>;SCHEMAS=schema1,schema2,schema3;'

                          Note: If you did not use passphrase to encrypt your private key, you can leave the "PASSPHRASE" empty.

                        - bigquery: /path_to_json_credential.json

                        For BigQuery, the credential file supports flexible authentication:

                        1. Application Default Credentials (ADC):
                           - Empty file: {} - uses environment authentication
                           - Override specific fields: {"project_id": "my-project", "datasets": [...]}

                          For more information about ADC, see https://cloud.google.com/docs/authentication/application-default-credentials

                        2. Explicit Service Account (bypasses ADC):
                           {
                             "project_id": "your-project-id",
                             "client_email": "your-client-email",
                             "private_key": "your-private-key",
                             "datasets": ["dataset_1", "dataset_2", ...]
                           }
                           Note: Both client_email and private_key must be provided together.

                        If "datasets" is not specified or is empty, all accessible datasets will be fetched.
  `).option("-o, --out-file <pathspec>", "compile all input files into a single files");
	showHelp(args);
	commander.default.parse(args);
	connectionHandler(commander.default);
}
exports.db2dbml = db2dbml;
exports.dbml2sql = dbml2sql;
exports.sql2dbml = sql2dbml;
